src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs

```diff
@@ -56,94 +68,44 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
-            if (!ifStatement.Parent.IsKind(SyntaxKind.Block))
-            {
-                return;
-            }
-
-            // We need to find the leftmost expression in the if-condition.  If this is a
-            // "x != null" expression, then we can replace it with "o is Type x".  
-            var condition = GetLeftmostCondition(ifStatement.Condition);
-            if (!condition.IsKind(SyntaxKind.NotEqualsExpression))
-            {
-                return;
-            }
-
-            // look for the form "x != null" or "null != x".
-            if (!IsNullCheckExpression(condition.Left, condition.Right) &&
-                !IsNullCheckExpression(condition.Right, condition.Left))
-            {
-                return;
-            }
-
-            var conditionName = condition.Left is IdentifierNameSyntax
-                ? (IdentifierNameSyntax)condition.Left
-                : (IdentifierNameSyntax)condition.Right;
-
-            // Now make sure the previous statement is "var a = ..."
-            var parentBlock = (BlockSyntax)ifStatement.Parent;
-            var ifIndex = parentBlock.Statements.IndexOf(ifStatement);
-            if (ifIndex == 0)
-            {
-                return;
-            }
-
-            var previousStatement = parentBlock.Statements[ifIndex - 1];
-            if (!previousStatement.IsKind(SyntaxKind.LocalDeclarationStatement))
+            var ifOrWhileStatement = (StatementSyntax)node;
+            
+            // if/while has to be in a block so we can at least look for a preceding local variable declaration.
+            if (!(ifOrWhileStatement.Parent is BlockSyntax parentBlock))
             {
                 return;
             }
 
-            var localDeclarationStatement = (LocalDeclarationStatementSyntax)previousStatement;
-            var variableDeclaration = localDeclarationStatement.Declaration;
-
-            if (variableDeclaration.Variables.Count != 1)
+            var leftmostCondition = GetLeftmostCondition(ifOrWhileStatement);
+            if (!leftmostCondition.IsKind(SyntaxKind.NotEqualsExpression))
             {
                 return;
             }
 
-            var declarator = variableDeclaration.Variables[0];
-            if (declarator.Initializer == null)
+            var comparison = (BinaryExpressionSyntax)leftmostCondition;
+            var operand = GetNullCheckOperand(comparison.Left, comparison.Right)?.WalkDownParentheses();
+            if (operand == null)
             {
                 return;
             }
 
-            if (!Equals(declarator.Identifier.ValueText, conditionName.Identifier.ValueText))
-            {
-                return;
-            }
-
-            // Make sure the initializer has the form "... = expr as Type;
-            var initializerValue = declarator.Initializer.Value;
-            if (!initializerValue.IsKind(SyntaxKind.AsExpression))
+            var blockStatements = parentBlock.Statements;
```

Comment:
@CyrusNajmabadi what is preferable to pass around here, `parentBlock` itself or `blockStatements`?

Since `BlockSyntax.Statements` returns a struct, ideally we should use `*/readonly*/ ref`, but I assumed that's ok for now